src/RollerPeriphery.sol

```diff
@@ -4,93 +4,159 @@ pragma solidity 0.8.13;
     error MinAssetsOrExcessError();
 
     /// @notice Redeem vault shares with slippage protection 
-    /// @param vault ERC4626 vault
+    /// @param roller AutoRoller vault
     /// @param shares Number of shares to redeem
     /// @param receiver Destination address for the returned assets
     /// @param minAmountOut Minimum amount of assets returned
     /// @return assets Amount of asset redeemable by the given number of shares
-    function redeem(ERC4626 vault, uint256 shares, address receiver, uint256 minAmountOut) external returns (uint256 assets) {
-        if ((assets = vault.redeem(shares, receiver, msg.sender)) < minAmountOut) {
+    function redeem(AutoRoller roller, uint256 shares, address receiver, uint256 minAmountOut) external returns (uint256 assets) {
+        if ((assets = roller.redeem(shares, receiver, msg.sender)) < minAmountOut) {
             revert MinAssetError();
         }
     }
 
+    function redeemForUnderlying(AutoRoller roller, uint256 shares, address receiver, uint256 minAmountOut) external returns (uint256 underlyingOut) {
+        AdapterLike adapter = AdapterLike(address(roller.adapter()));
+
+        if ((underlyingOut = adapter.unwrapTarget(roller.redeem(shares, address(this), msg.sender))) < minAmountOut) {
+            revert MinUnderlyingError();
+        }
+
+        ERC20(adapter.underlying()).safeTransfer(receiver, underlyingOut);
+    }
+
     /// @notice Withdraw underlying asset from vault with slippage protection 
-    /// @param vault ERC4626 vault
+    /// @param roller AutoRoller vault
     /// @param assets Amount of asset requested for withdrawal
     /// @param receiver Destination address for the returned assets
     /// @param maxSharesOut Maximum amount of shares burned
     /// @return shares Number of shares to redeem
-    function withdraw(ERC4626 vault, uint256 assets, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
-        if ((shares = vault.withdraw(assets, receiver, msg.sender)) > maxSharesOut) {
+    function withdraw(AutoRoller roller, uint256 assets, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
+        if ((shares = roller.withdraw(assets, receiver, msg.sender)) > maxSharesOut) {
+            revert MaxSharesError();
+        }
+    }
+
+    function withdrawUnderlying(AutoRoller roller, uint256 underlyingOut, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
+        AdapterLike adapter = AdapterLike(address(roller.adapter()));
+
+        uint256 assetOut = underlyingOut.divWadDown(adapter.scale()); // Tokens out, round down.
+
+        if ((shares = roller.withdraw(assetOut, address(this), msg.sender)) > maxSharesOut) {
             revert MaxSharesError();
         }
+
+        ERC20(adapter.underlying()).safeTransfer(
+            receiver,
+            adapter.unwrapTarget(roller.asset().balanceOf(address(this)))
```

Comment:
Stylistic suggestion to pull out the argument into its own line for greater readability. 
E.g (note I've assumed the `round down` operation takes away some dust amount away from `underlyingOut`) : 
```
uint256 underlyingOutSubDust = adapter.unwrapTarget(roller.asset().balanceOf(address(this));

ERC20(adapter.underlying()).safeTransfer(receiver, underlyingOutSubDust);
```

Reply:
👍